File: code/Python/Chapter11/11-2/knapsack.py
class Item:
    def __init__(self, weight, value):
        self.weight = weight
        self.value = value
        self.unit_value = value / weight
    
    def __str__(self):
        return "Item(weight=%d, value=%d, unit_value=%.2f)" % (
            self.weight, self.value, self.unit_value
        )

def knapsack(items, capacity):
    items.sort(key=lambda x: x.unit_value, reverse=True)

    selected_items = []
    selected_weight = 0

    for item in items:
        if selected_weight + item.weight <= capacity:
            selected_weight += item.weight
            selected_items.append(item)
        else:
            portion = (capacity - selected_weight) / item.weight
            item.weight *= portion
            item.value *= portion
            selected_items.append(item)
            break

    return selected_items

items = [
    Item(10, 20),
    Item(5, 30), 
    Item(15, 15),
    Item(10, 25),
    Item(20, 10)
]
capacity = 30

selected_items = knapsack(items, capacity)
for item in selected_items:
    print(item)

File: code/Java/Chapter11/11-2/Knapsack.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class Knapsack {
    public static class Item implements Comparable<Item> {
        double weight;
        double value;
        double unitValue;

        public Item(double weight, double value) {
            this.weight = weight;
            this.value = value;
            this.unitValue = value / weight;
        }

        @Override
        public int compareTo(Item obj) {
            return Double.valueOf(this.unitValue).compareTo(obj.unitValue);
        }

        @Override
        public String toString() {
            return "Item(weight=" + weight + ", value=" + value + ", unitValue=" + unitValue + ")";
        }
    }

    public static Item[] knapsack(Item[] items, int capacity) {
        Arrays.sort(items, Collections.reverseOrder());

        ArrayList<Item> selectedItems = new ArrayList<Item>();
        double selectedWeight = 0;

        for (Item item : items) {
            if (selectedWeight + item.weight <= capacity) {
                selectedWeight += item.weight;
                selectedItems.add(item);
            } else {
                double portion = (capacity - selectedWeight) / item.weight;
                item.weight *= portion;
                item.value *= portion;
                selectedItems.add(item);
                break;
            }
        }

        return selectedItems.toArray(new Item[selectedItems.size()]);
    }

    public static void main(String[] args) {
        Item[] items = {
                new Item(10, 20),
                new Item(5, 30),
                new Item(15, 15),
                new Item(10, 25),
                new Item(20, 10)
        };
        int capacity = 30;

        Item[] selectedItems = knapsack(items, capacity);
        for (Item item : selectedItems) {
            System.out.println(item);
        }
    }
}
